cmd/gazctl/gazctlcmd/shards_prune.go

```diff
@@ -53,29 +52,40 @@ func (cmd *cmdShardsPrune) Execute([]string) error {
 
 	for _, shard := range listShards(rsc, cmd.Selector).Shards {
 		metrics.shardsTotal++
-		var lastHints = fetchOldestHints(ctx, rsc, shard.Spec.Id)
+
+		var allHints, err = consumer.FetchHints(ctx, rsc, &pc.GetHintsRequest{
+			Shard: shard.Spec.Id,
+		})
+		mbp.Must(err, "failed to fetch hints")
+
 		var recoveryLog = shard.Spec.RecoveryLog()
 
-		// We require that we see hints for _all_ shards before we may make _any_ deletions.
-		// This is because shards could technically include segments from any log,
-		// and without comprehensive hints which are proof-positive that _no_ shard
-		// references a given journal fragment, we cannot be sure it's safe to remove.
-		// For this reason, we must track the journals to be skipped, so we can be sure
-		// we don't prune journals that are used by a shard that hasn't persisted hints.
-		if lastHints != nil && len(lastHints.LiveNodes) > 0 {
-			foldHintsIntoSegments(*lastHints, logSegmentSets)
-		} else {
-			skipRecoveryLogs[recoveryLog] = true
-			metrics.skippedJournals++
-			var reason = "has not written backup hints required for pruning"
-			if lastHints != nil {
-				reason = "hints have no live files"
+		for _, curHints := range append(allHints.BackupHints, allHints.PrimaryHints) {
+			var hints = curHints.Hints
+
+			// We require that we see _all_ hints for a shards before we may make _any_ deletions.
+			// This is because shards could technically include segments from any log,
+			// and without comprehensive hints which are proof-positive that _no_ shard
+			// references a given journal fragment, we cannot be sure it's safe to remove.
+			// For this reason, we must track the journals to be skipped, so we can be sure
+			// we don't prune journals that are used by a shard that hasn't persisted hints.
+			if hints != nil && len(hints.LiveNodes) > 0 {
+				foldHintsIntoSegments(*hints, logSegmentSets)
+			} else {
+				skipRecoveryLogs[recoveryLog] = true
```

Comment:
Not something that's changed here, but this just made me curious. If there's non-nil hints, but they don't have any `LiveNodes`, could that indicate that all the files have been deleted, and thus we _should_ prune that journal still? Not proposing that we handle that corner case, but more just checking my understanding.

Reply:
Hmm, that sounds conceptually right but I don't recall if hinted log replay can function if there are literally no LiveNodes. Possibly it handles it by jumping to the head of the log, or possibly it errors out because it wouldn't have an `(offset, sequenceNo, checksum)` tuple to resume reading from.

But if all data is also deleted, that can also be problematic because there must be written & persisted data that's old enough for the reader's offset to be eligible to jump to the log head.

(To my knowledge there's no actual scenario where `LiveNodes` is intentionally empty.)